odoo/addons/base/models/res_company.py

```diff
@@ -376,9 +376,16 @@ def _get_main_company(self):
     def __accessible_branches(self):
         # Get branches of this company that the current user can use
         self.ensure_one()
-        children = self.search([('id', 'child_of', self.id)])
-        accessible = self.env.companies
-        return (children & accessible).ids
+
+        rslt = []
+        to_treat = [company for company in self.env.companies.filtered(lambda x: not x.parent_id)]
+        while to_treat:
+            company = to_treat.pop(0)
+            if company in self.env.companies:
+                rslt.append(company.id)
+                to_treat = [child for child in company.child_ids] + to_treat
```

Comment:
We only want the children of `self`, which can only contain one record (see `ensure_one` above)
```suggestion
        current = self  # might need sudo?
        accessible = self.env.companies
        while current:
            rslt.extend((current & accessible).ids)
            current = current.child_ids
```